### DISCUSSION TITLE
Help：Agent NEVER get any result from workflow tool

### DISCUSSION THREAD

**Discussion by Zerglingzl** [langgenius/dify#12107] [Help]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

1, I created a Agent, and add a workflow in "tool" area(enabled for sure), let's name it Workflow_Tool.
2, I asked this Agent to use Workflow_Tool to address content issue, the Agent did call the tool.
3, BUT, when the tool is done, the Agent always say that the Workflow_Tool returned NULL {}, which is impossible.
4, I checked the workflow run history. history clearly shows that workflow received the task, finished it, and RETURNED result.

So, why the Agent can't handle the return value of a workflow tool? is there any bugs of default API definition? Please help, thanks a lot!

**Comment by user905** [comment]
@dosu - this is still an issue in Dify 1.2.0. My custom tool works when run as its own node in a workflow, but when run by an agent, nothing is returned. 

This is the tool response when run by an Agent: 
`
{
  "output": {
    "llm_response": "",
    "tool_responses": [
      {
        "tool_call_id": "call_O15ZwDJSMLRfypwwDBmwzfNl",
        "tool_call_input": {},
        "tool_call_name": "projects_get",
        "tool_response": ""
      }
    ]
  }
}
`

This is the response when the tool is run as its own node: 
`
{
  "text": "{\"projects\": [{\"id\": 389, \"name\": \"Salt Waste Processing Facility\", \"acronym\": \"SWPF\", \"siteCode\": \"SRS\", \"program\": \"EM\"} ]}",
  "files": [],
  "json": []
}
`

This is the output when running via an agent. Note that it is successful:
![image](https://github.com/user-attachments/assets/9f7e9c8b-a2e0-49a0-9c5f-aade3b10cfaf)

The full input is: 
`
{
  "model": {
    "provider": "langgenius/azure_openai/azure_openai",
    "model": "gpt-4o",
    "model_type": "llm",
    "mode": "chat",
    "completion_params": {
      "temperature": 0.5,
      "top_p": 0.85,
      "presence_penalty": 0.2,
      "frequency_penalty": 0.3
    },
    "type": "model-selector"
  },
  "tools": [
    {
      "provider_name": "682a753b-74e4-4434-b86c-4261aac96680",
      "type": "api",
      "tool_name": "projects_get",
      "tool_label": "projects_get",
      "settings": {},
      "parameters": {},
      "enabled": true,
      "extra": {
        "description": "Returns a list of PARS projects with basic information for each."
      }
    }
  ],
  "instruction": "when a user inputs a question, get the list of projects and determine the PARSID/ID of the specific project. If the projects_get tool returns nothing say, 'No projects returned.' Otherwise, if you aren't sure, say \"I don't know\". ",
  "query": "wtp"
}
`

**Comment by linshiyan1992** [comment]
Hello, any update on this issue?  Do I need to do an upgrade to solve this?  @dosu

**Reply by user905** [reply]
I haven't gotten this working yet.
